features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/GatewayEnvironments/AddEditGWEnvironment.jsx

```diff
@@ -58,48 +64,73 @@ function reducer(state, { field, value }) {
 }
 
 /**
- * Render a pop-up dialog to add/edit an Gateway label
+ * Render a pop-up dialog to add/edit a Gateway Environment
  * @param {JSON} props .
  * @returns {JSX}.
  */
-function AddEditMGLabel(props) {
+function AddEditGWEnvironment(props) {
     const {
         updateList, dataRow, icon, triggerButtonText, title,
     } = props;
     const classes = useStyles();
 
+    const defaultVhost = {
+        host: '', httpContext: '', httpsPort: 443, httpPort: 80, wssPort: 8099, wsPort: 9099, isNew: true,
+    };
     const [initialState, setInitialState] = useState({
+        displayName: '',
         description: '',
-        hosts: [],
+        vhosts: [defaultVhost],
     });
     const [editMode, setIsEditMode] = useState(false);
 
     const [state, dispatch] = useReducer(reducer, initialState);
-    const { name, description, hosts } = state;
+    const {
+        name, displayName, description, vhosts,
+    } = state;
 
     const onChange = (e) => {
         dispatch({ field: e.target.name, value: e.target.value });
     };
     useEffect(() => {
         setInitialState({
+            displayName: '',
             description: '',
-            hosts: [],
+            vhosts: [defaultVhost],
         });
     }, []);
 
-    const handleHostValidation = (hostName) => {
-        if (hostName === undefined) {
+    const handleHostValidation = (vhost) => {
+        if (vhost === undefined) {
             return false;
         }
-        const schema = Joi.string().uri().empty();
-        const validationError = schema.validate(hostName).error;
+        if (!vhost.host) {
+            return 'Host of Vhost is empty';
+        }
+        // same pattern used in admin Rest API
+        const hostPattern = '^(((\\*[\\.-])?(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*'
+            + '([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\\-]*[A-Za-z0-9]))|((([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*'
+            + '[a-zA-Z0-9])\\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\\-]*[A-Za-z0-9]))([\\.-]\\*)?|\\*)$';
+        const hostRegex = new RegExp(hostPattern, 'g');
+        const validHost = vhost.host && vhost.host.match(hostRegex);
+        if (!validHost) {
+            return 'Invalid Host';
+        }
 
-        if (validationError) {
-            const errorType = validationError.details[0].type;
-            if (errorType === 'any.empty') {
-                return 'Host is empty';
-            } else if (errorType === 'string.uri') {
-                return 'Invalid Host';
+        // same pattern used in admin Rest API
+        const httpContextRegex = /^\/?([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])*$/g;
+        // empty http context are valid
+        const validHttpContext = !vhost.httpContext || vhost.httpContext.match(httpContextRegex);
+        if (!validHttpContext) {
+            return 'Invalid Http context';
+        }
+
+        let portError;
+        const ports = ['httpPort', 'httpsPort', 'wsPort', 'wssPort'];
```

Comment:
Can't we get this list from OAS definition ?

I think these are constant config keys, at any point if we change these config keys in deployment.toml we needed to change these strings here , So better to get this from OAS definition if possible

Reply:
This array contains the properties of the object that gets from the response of the Rest API. Instead of using `vhost.httpPort` and `vhost.httpsPort` I used this looping since it is not duplicating codes. The values read from deployment toml is convert to this object and returns from the Rest API.